Add `AF_PACKET` support

mertyildiran opened the pull request:
This PR adds `AF_PACKET` support as the capture source. `AF_PACKET` is supported since Linux kernel version 2.2

Adds `AF_XDP` support as the capture source. `AF_XDP` is supported since Linux kernel version 4.18

Adds `packet-capture` config alongside `tls` config to set the capture source. Its default value is `libpcap` so Mizu uses `libpcap` as before by default. The possible values of `packet-capture` flag are;

- `af_xdp`
- `af_packet`
- `libpcap`

an input that's not valid falls back to `libpcap` without a panic.

`AF_PACKET` has a lower drop rate when it's compared to `libpcap`.

mertyildiran commented:
Maybe make `api.Pcap` and `api.Afpacket` distinction to differentiate the capture source.

mertyildiran commented:
The base images for the cross-compilation build targets below might need `linux-headers` dependency for `AF_PACKET` support:
- `--build-arg BUILDARCH=amd64 --build-arg TARGETARCH=arm64v8`
- `--build-arg BUILDARCH=arm64v8 --build-arg TARGETARCH=amd64`

So, merge https://github.com/up9inc/mizu/pull/910 first then check those build targets, only then merge this PR.

nimrod-up9 commented:
I'm unfamiliar with AF_PACKET and PF_PACKET.
I googled a bit and I couldn't find an explanation of what the AF_PACKET and PF_PACKET families are.
The [wiki on Berkley sockets](https://en.wikipedia.org/wiki/Berkeley_sockets) does not mention AF_PACKET (only AF_INTET).
The [man page for packet](https://man7.org/linux/man-pages/man7/packet.7.html) only mentions AF_PACKET as the only option.
Found a short comparison in [Elastic forums](https://discuss.elastic.co/t/performance-difference-between-af-packet-libpcap/69766).
Could you please explain or link to a description of AF_PACKET and PF_PACKET?

mertyildiran commented on tap/settings.go:
```diff
@@ -14,8 +14,8 @@ const (
 	MemoryProfilingTimeIntervalSeconds             = "MEMORY_PROFILING_TIME_INTERVAL"
 	MaxBufferedPagesTotalEnvVarName                = "MAX_BUFFERED_PAGES_TOTAL"
 	MaxBufferedPagesPerConnectionEnvVarName        = "MAX_BUFFERED_PAGES_PER_CONNECTION"
-	MaxBufferedPagesTotalDefaultValue              = 5000
-	MaxBufferedPagesPerConnectionDefaultValue      = 5000
```
Should we set these values `0`? `0` is the default value inside the `gopacket`.

mertyildiran replied:
Let's not change these consts with this PR. Reverted with https://github.com/up9inc/mizu/commit/3f6c98db13ba4d873562c9e3ba9f1c3239caa2a5

mertyildiran commented on tap/passive_tapper.go:
```diff
@@ -187,7 +189,7 @@ func initializePassiveTapper(opts *TapOpts, outputItems chan *api.OutputChannelI
 	diagnose.InitializeErrorsMap(*debug, *verbose, *quiet)
 	diagnose.InitializeTapperInternalStats()
 
-	mainPacketInputChan = make(chan source.TcpPacketInfo)
```
What should be the buffer size?

mertyildiran commented:
> I'm unfamiliar with AF_PACKET and PF_PACKET. I googled a bit and I couldn't find an explanation of what the AF_PACKET and PF_PACKET families are. The [wiki on Berkley sockets](https://en.wikipedia.org/wiki/Berkeley_sockets) does not mention AF_PACKET (only AF_INTET). The [man page for packet](https://man7.org/linux/man-pages/man7/packet.7.html) only mentions AF_PACKET as the only option. Found a short comparison in [Elastic forums](https://discuss.elastic.co/t/performance-difference-between-af-packet-libpcap/69766). Could you please explain or link to a description of AF_PACKET and PF_PACKET?

### GoPacket

@nimrod-up9 it's about being [zero-copy](https://en.wikipedia.org/wiki/Zero-copy) and `gopacket`-`libpcap` implementation's cgo calls performance drawback. See the author's comments:
- https://github.com/google/gopacket/issues/329#issuecomment-317471084
- https://github.com/gconnell/stenographer/blob/master/DESIGN.md#packet-sniffingwriting

`libpcap` implementation is not zero-copy because of this line: https://github.com/google/gopacket/blob/v1.1.19/pcap/pcap.go#L255

`AF_PACKET` and `PF_PACKET` are essentially the same thing;
- https://stackoverflow.com/questions/6729366/what-is-the-difference-between-af-inet-and-pf-inet-in-socket-programming
- https://github.com/torvalds/linux/blob/169e77764adc041b1dacba84ea90516a895d43b2/include/linux/socket.h#L252

It's just in `gopacket` they named that implementation `AF_PACKET` because of;
- https://github.com/google/gopacket/blob/v1.1.19/afpacket/afpacket.go#L239

So, [`afpacket` package](https://github.com/google/gopacket/tree/v1.1.19/afpacket) is an implementation that's more native to Go language and achieves [zero-copy](https://en.wikipedia.org/wiki/Zero-copy). Which is why it performs better.

### Linux

Raw Linux sockets (`SOCK_RAW`) operates on the data link layer and just before passing it to the network layer, it passes to the our raw socket as well;

![2014_10_08_linux_pf_packet](https://user-images.githubusercontent.com/2502080/167289824-db6bd26a-4942-474e-9f2c-a03433e432a0.jpg)

I tracked the `socket` syscall;
- https://github.com/torvalds/linux/blob/30c8e80f79329617012f07b09b70114592092ea4/net/socket.c#L1541
- https://github.com/torvalds/linux/blob/30c8e80f79329617012f07b09b70114592092ea4/net/socket.c#L1397
- https://github.com/torvalds/linux/blob/84bfcc0b6994057905cf98d2c5cedef48b3322b5/security/apparmor/lsm.c#L835
- https://github.com/torvalds/linux/blob/5bfc75d92efd494db37f5c4c173d3639d4772966/security/apparmor/net.c#L134
- https://github.com/torvalds/linux/blob/5bfc75d92efd494db37f5c4c173d3639d4772966/security/apparmor/net.c#L108

It's hard to understand from that point without being a kernel developer, AppArmor is complicated ¯\\\_(ツ)\_/¯

But at least I can tell to that point, I didn't see an `AF_` vs. `PF_` distinction in the kernel, related to socket creation and [this line](https://github.com/torvalds/linux/blob/169e77764adc041b1dacba84ea90516a895d43b2/include/linux/socket.h#L252) tells us they are the same thing.

mertyildiran commented on tap/passive_tapper.go:
```diff
@@ -187,7 +189,7 @@ func initializePassiveTapper(opts *TapOpts, outputItems chan *api.OutputChannelI
 	diagnose.InitializeTapperInternalStats()
 
-	mainPacketInputChan = make(chan source.TcpPacketInfo)
+	mainPacketInputChan = make(chan source.TcpPacketInfo, 1000)
```
In `gopacket` they use `1000` as the channel buffer size in here:

https://github.com/google/gopacket/blob/v1.1.19/packet.go#L860

So, I guess `1000` is OK?

mertyildiran commented:
@nimrod-up9 FYI: I realized that I broke the code while fixing a linter error. Fixed it with https://github.com/up9inc/mizu/commit/df66231ed5da8b11765b5aeb3ba555a0ca0fec32 Please try again if it didn't work for the last time.

mertyildiran commented:
# How Well AF_PACKET Improves the Performance?

Download [smallFlows.pcap](https://s3.amazonaws.com/tcpreplay-pcap-files/smallFlows.pcap)

Running;

```
sudo tcpreplay -i enp6s0 -tK --loop 1000 --unique-ip smallFlows.pcap
```

generates ~120MB/s traffic on my machine. (I'm only that much able to ramp up the traffic on my machine)

NIC: Intel® Ethernet Controller I211

## CONSISTENCY

Pay attention to `matchedPairs` values and `ERROR ▶ error converting message to json &{{http Hyperte...` logs.
The error message is caused by the packet loss.

`AF_PACKET` has 4742 matched pairs.

`libpcap` has 3151 matched pairs.

So, I guess `AF_PACKET` is more consistent for tracking a connection when it's compared to `libpcap`.
Although the results might be accidental as well.

### libpcap

```
[2022-05-08T15:38:00.717+0300] INFO  ▶ Now tapping:  ▶ [864525 passive_tapper.go:116 printNewTapTargets]
[2022-05-08T15:38:00.717+0300] WARNING ▶ Received socket message of type startTime for which no handlers are defined ▶ [864525 main.go:366 handleIncomingMessageAsTapper]
[2022-05-08T15:38:37.181+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013517129 0xc00199d4a0 0xc000eef030 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:38:37.181+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013517129 0xc00199d4a0 0xc000eef030 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:38:40.503+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013520445 0xc0057f5f40 0xc0003b7f10 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:38:40.503+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013520445 0xc0057f5f40 0xc0003b7f10 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:38:46.965+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013526902 0xc004d89180 0xc0090b7f10 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:38:46.965+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013526902 0xc004d89180 0xc0090b7f10 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:39:00.715+0300] INFO  ▶ 1m0.000944902s (errors: 3589, errTypes:2) - Errors Summary: map[FSM-rejection:637 OptionChecker-rejection:2952] ▶ [864525 passive_tapper.go:132 printPeriodicStats]
[2022-05-08T15:39:00.715+0300] INFO  ▶ mem: 74301288, goroutines: 2431 ▶ [864525 passive_tapper.go:142 printPeriodicStats]
[2022-05-08T15:39:00.715+0300] INFO  ▶ cleaner - flushed connections: 0, closed connections: 0, deleted messages: 0 ▶ [864525 passive_tapper.go:150 printPeriodicStats]
[2022-05-08T15:39:00.715+0300] INFO  ▶ app stats - {"processedBytes":80180457,"packetsCount":86455,"tcpPacketsCount":80390,"reassembledTcpPayloadsCount":36916,"tlsConnectionsCount":0,"matchedPairs":2225,"droppedTcpStreams":2148} ▶ [864525 passive_tapper.go:158 printPeriodicStats]
[2022-05-08T15:39:03.835+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013543812 0xc00c0d5810 0xc008e50000 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:39:03.835+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013543812 0xc00c0d5810 0xc008e50000 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:00.714+0300] INFO  ▶ 2m0.000687836s (errors: 4662, errTypes:2) - Errors Summary: map[FSM-rejection:832 OptionChecker-rejection:3830] ▶ [864525 passive_tapper.go:132 printPeriodicStats]
[2022-05-08T15:40:00.715+0300] INFO  ▶ mem: 91878376, goroutines: 3313 ▶ [864525 passive_tapper.go:142 printPeriodicStats]
[2022-05-08T15:40:00.715+0300] INFO  ▶ cleaner - flushed connections: 0, closed connections: 0, deleted messages: 0 ▶ [864525 passive_tapper.go:150 printPeriodicStats]
[2022-05-08T15:40:00.715+0300] INFO  ▶ app stats - {"processedBytes":68194334,"packetsCount":66052,"tcpPacketsCount":32298,"reassembledTcpPayloadsCount":16024,"tlsConnectionsCount":0,"matchedPairs":3019,"droppedTcpStreams":943} ▶ [864525 passive_tapper.go:158 printPeriodicStats]
[2022-05-08T15:40:20.725+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013491125 0xc002882a00 0xc005b62fc0 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:20.725+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013491125 0xc002882a00 0xc005b62fc0 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:20.727+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013494338 0xc003d6c050 0xc009f68930 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:20.727+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013494338 0xc003d6c050 0xc009f68930 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:20.736+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013491125 0xc005a58ff0 0xc0053a0cb0 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:20.736+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013491125 0xc005a58ff0 0xc0053a0cb0 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:30.741+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013500907 0xc0015ae8c0 0xc0057af880 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:30.741+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013500907 0xc0015ae8c0 0xc0057af880 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:30.743+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013500722 0xc005a598b0 0xc0086b4070 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:30.743+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013500722 0xc005a598b0 0xc0086b4070 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:30.758+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013500862 0xc0017b3950 0xc000f685b0 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:30.758+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013500862 0xc0017b3950 0xc000f685b0 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:40.723+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013510741 0xc001bba000 0xc0003b72d0 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:40.723+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013510741 0xc001bba000 0xc0003b72d0 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:40.726+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013517168 0xc00438a640 0xc0002908c0 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:40.726+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013517168 0xc00438a640 0xc0002908c0 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:40.730+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013517129 0xc0142afc20 0xc000289f80 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:40.730+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013517129 0xc0142afc20 0xc000289f80 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:50.731+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013530234 0xc0064ccb40 0xc0002908c0 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:50.731+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013530234 0xc0064ccb40 0xc0002908c0 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:50.737+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013526918 0xc0034f3b80 0xc009f68930 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:40:50.737+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013526918 0xc0034f3b80 0xc009f68930 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:41:00.714+0300] INFO  ▶ 3m0.000606925s (errors: 4989, errTypes:2) - Errors Summary: map[FSM-rejection:836 OptionChecker-rejection:4153] ▶ [864525 passive_tapper.go:132 printPeriodicStats]
[2022-05-08T15:41:00.715+0300] INFO  ▶ mem: 78361048, goroutines: 1529 ▶ [864525 passive_tapper.go:142 printPeriodicStats]
[2022-05-08T15:41:00.715+0300] INFO  ▶ cleaner - flushed connections: 503, closed connections: 4030, deleted messages: 0 ▶ [864525 passive_tapper.go:150 printPeriodicStats]
[2022-05-08T15:41:00.715+0300] INFO  ▶ app stats - {"processedBytes":95615986,"packetsCount":62469,"tcpPacketsCount":3557,"reassembledTcpPayloadsCount":2075,"tlsConnectionsCount":0,"matchedPairs":3151,"droppedTcpStreams":5} ▶ [864525 passive_tapper.go:158 printPeriodicStats]
[2022-05-08T15:41:00.725+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013537165 0xc0031434a0 0xc0002602a0 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:41:00.725+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013537165 0xc0031434a0 0xc0002602a0 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:41:00.725+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013537113 0xc00438a640 0xc0001daa10 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:41:00.725+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013537113 0xc00438a640 0xc0001daa10 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:41:00.726+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013533607 0xc0017b3900 0xc000356690 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:41:00.726+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013533607 0xc0017b3900 0xc000356690 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:41:00.726+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013533800 0xc001bba640 0xc0005a65b0 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:41:00.726+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013533800 0xc001bba640 0xc0005a65b0 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:41:00.728+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013537113 0xc003d96500 0xc00040f260 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:41:00.728+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013537113 0xc003d96500 0xc00040f260 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:41:00.770+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013533572 0xc0015ae870 0xc00c64bdc0 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:41:00.770+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013533572 0xc0015ae870 0xc00c64bdc0 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:41:00.772+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013537264 0xc00ab81a90 0xc0086b4770 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
[2022-05-08T15:41:00.772+0300] ERROR ▶ error converting message to json &{{http Hypertext Transfer Protocol -- HTTP/1.1 HTTP http 1.1 #205cf5 #ffffff 12 https://datatracker.ietf.org/doc/html/rfc2616 [80 443 8080] 0} pcap 1652013537264 0xc00ab81a90 0xc0086b4770 <nil> }, err: json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR, (json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR,json: error calling MarshalJSON for type api.HTTPPayload: Failed converting response to HAR) ▶ [864525 main.go:282 pipeTapChannelToSocket]
^C[2022-05-08T15:41:05.496+0300] INFO  ▶ Shutting down... ▶ [864494 utils.go:39 func1]
[2022-05-08T15:41:05.496+0300] INFO  ▶ Exiting ▶ [864525 main.go:82 main]
```

### AF_PACKET

```
[2022-05-08T15:34:12.409+0300] INFO  ▶ Now tapping:  ▶ [863308 passive_tapper.go:117 printNewTapTargets]
[2022-05-08T15:34:12.409+0300] WARNING ▶ Received socket message of type startTime for which no handlers are defined ▶ [863308 main.go:366 handleIncomingMessageAsTapper]
[2022-05-08T15:35:12.407+0300] INFO  ▶ 1m0.000998371s (errors: 2519, errTypes:2) - Errors Summary: map[FSM-rejection:174 OptionChecker-rejection:2345] ▶ [863308 passive_tapper.go:133 printPeriodicStats]
[2022-05-08T15:35:12.408+0300] INFO  ▶ mem: 150729208, goroutines: 3419 ▶ [863308 passive_tapper.go:143 printPeriodicStats]
[2022-05-08T15:35:12.408+0300] INFO  ▶ cleaner - flushed connections: 0, closed connections: 0, deleted messages: 0 ▶ [863308 passive_tapper.go:151 printPeriodicStats]
[2022-05-08T15:35:12.412+0300] INFO  ▶ app stats - {"processedBytes":145401760,"packetsCount":109890,"tcpPacketsCount":104495,"reassembledTcpPayloadsCount":50019,"tlsConnectionsCount":0,"matchedPairs":3440,"droppedTcpStreams":5374} ▶ [863308 passive_tapper.go:159 printPeriodicStats]
[2022-05-08T15:36:12.406+0300] INFO  ▶ 2m0.000510387s (errors: 3999, errTypes:2) - Errors Summary: map[FSM-rejection:286 OptionChecker-rejection:3713] ▶ [863308 passive_tapper.go:133 printPeriodicStats]
[2022-05-08T15:36:12.407+0300] INFO  ▶ mem: 110530696, goroutines: 2357 ▶ [863308 passive_tapper.go:143 printPeriodicStats]
[2022-05-08T15:36:12.407+0300] INFO  ▶ cleaner - flushed connections: 0, closed connections: 0, deleted messages: 0 ▶ [863308 passive_tapper.go:151 printPeriodicStats]
[2022-05-08T15:36:12.407+0300] INFO  ▶ app stats - {"processedBytes":146028165,"packetsCount":102080,"tcpPacketsCount":48290,"reassembledTcpPayloadsCount":21494,"tlsConnectionsCount":0,"matchedPairs":4742,"droppedTcpStreams":3822} ▶ [863308 passive_tapper.go:159 printPeriodicStats]
^C[2022-05-08T15:37:02.104+0300] INFO  ▶ Exiting ▶ [863308 main.go:82 main]
[2022-05-08T15:37:02.104+0300] INFO  ▶ Shutting down... ▶ [863272 utils.go:39 func1]
```

## DROP RATE

Same Tcpreplay test;

```
sudo tcpreplay -i enp6s0 -tK --loop 1000 --unique-ip smallFlows.pcap
```

libpcap has ~0.504 drop rate.

AF_PACKET has ~0.495 drop rate.

Means that AF_PACKET provides 1% capture performance improvement.

### libpcap

```go
go func() {
    ticker := time.NewTicker(time.Second)

    for {
        <-ticker.C
        stats, err := result.handle.Stats()
        if err != nil {
            fmt.Printf("OOPS!\n")
        }
        fmt.Printf("stats: %+v\n", stats)
        dropRate := float64(stats.PacketsReceived) / float64(stats.PacketsReceived+stats.PacketsDropped)
        fmt.Printf("dropRate: %v\n", dropRate)
    }
}()
```

```
[2022-05-08T16:03:01.018+0300] INFO  ▶ Now tapping:  ▶ [873658 passive_tapper.go:116 printNewTapTargets]
[2022-05-08T16:03:01.018+0300] WARNING ▶ Received socket message of type startTime for which no handlers are defined ▶ [873658 main.go:366 handleIncomingMessageAsTapper]
stats: &{PacketsReceived:3808 PacketsDropped:0 PacketsIfDropped:0}
dropRate: 1
stats: &{PacketsReceived:75656 PacketsDropped:60849 PacketsIfDropped:0}
dropRate: 0.5542361085674518
stats: &{PacketsReceived:261720 PacketsDropped:246913 PacketsIfDropped:0}
dropRate: 0.514555681601469
stats: &{PacketsReceived:447563 PacketsDropped:432756 PacketsIfDropped:0}
dropRate: 0.5084100195497314
stats: &{PacketsReceived:634733 PacketsDropped:618396 PacketsIfDropped:0}
dropRate: 0.5065184829335209
stats: &{PacketsReceived:827017 PacketsDropped:803775 PacketsIfDropped:0}
dropRate: 0.5071259854107697
stats: &{PacketsReceived:1012919 PacketsDropped:989677 PacketsIfDropped:0}
dropRate: 0.5058029677478633
stats: &{PacketsReceived:1198869 PacketsDropped:1175627 PacketsIfDropped:0}
dropRate: 0.5048940912092503
stats: &{PacketsReceived:1385310 PacketsDropped:1362068 PacketsIfDropped:0}
dropRate: 0.5042298511526262
stats: &{PacketsReceived:1571265 PacketsDropped:1548023 PacketsIfDropped:0}
dropRate: 0.503725529672156
stats: &{PacketsReceived:1760659 PacketsDropped:1737417 PacketsIfDropped:0}
dropRate: 0.5033221119266705
stats: &{PacketsReceived:1950186 PacketsDropped:1922036 PacketsIfDropped:0}
dropRate: 0.5036348639101788
stats: &{PacketsReceived:2136049 PacketsDropped:2107899 PacketsIfDropped:0}
dropRate: 0.5033164873839171
stats: &{PacketsReceived:2321936 PacketsDropped:2293786 PacketsIfDropped:0}
dropRate: 0.5030493604250863
stats: &{PacketsReceived:2508142 PacketsDropped:2479992 PacketsIfDropped:0}
dropRate: 0.5028216964500152
stats: &{PacketsReceived:2697601 PacketsDropped:2669451 PacketsIfDropped:0}
dropRate: 0.5026224825099515
stats: &{PacketsReceived:2883663 PacketsDropped:2855513 PacketsIfDropped:0}
dropRate: 0.5024524426503038
stats: &{PacketsReceived:3069933 PacketsDropped:3039837 PacketsIfDropped:0}
dropRate: 0.5024629405034887
stats: &{PacketsReceived:3258584 PacketsDropped:3226829 PacketsIfDropped:0}
dropRate: 0.5024481864146508
stats: &{PacketsReceived:3447402 PacketsDropped:3415647 PacketsIfDropped:0}
dropRate: 0.5023134761240958
^C[2022-05-08T16:03:21.603+0300] INFO  ▶ Exiting ▶ [873658 main.go:82 main]
[2022-05-08T16:03:21.603+0300] INFO  ▶ Shutting down... ▶ [873626 utils.go:39 func1]
```

### AF_PACKET

```go
go func() {
    ticker := time.NewTicker(time.Second)

    for {
        <-ticker.C
        _, stats, err := capture.SocketStats()
        if err != nil {
            fmt.Printf("OOPS!\n")
        }
        fmt.Printf("stats: %+v\n", stats)
        dropRate := float64(stats.Drops()) / float64(stats.Drops()+stats.Packets())
        fmt.Printf("dropRate: %v\n", dropRate)
    }
}()
```

```
[2022-05-08T16:05:12.995+0300] INFO  ▶ Now tapping:  ▶ [875154 passive_tapper.go:117 printNewTapTargets]
[2022-05-08T16:05:12.995+0300] WARNING ▶ Received socket message of type startTime for which no handlers are defined ▶ [875154 main.go:366 handleIncomingMessageAsTapper]
stats: {tp_packets:5183 tp_drops:176 tp_freeze_q_cnt:14}
dropRate: 0.03284194812465012
stats: {tp_packets:91281 tp_drops:75128 tp_freeze_q_cnt:24}
dropRate: 0.4514659663840297
stats: {tp_packets:278343 tp_drops:262190 tp_freeze_q_cnt:24}
dropRate: 0.48505826656281853
stats: {tp_packets:465659 tp_drops:449506 tp_freeze_q_cnt:24}
dropRate: 0.4911748154704343
stats: {tp_packets:653784 tp_drops:637631 tp_freeze_q_cnt:24}
dropRate: 0.493746007286581
stats: {tp_packets:847264 tp_drops:826436 tp_freeze_q_cnt:25}
dropRate: 0.49377785744159647
stats: {tp_packets:1037668 tp_drops:1016840 tp_freeze_q_cnt:25}
dropRate: 0.49493114653240583
stats: {tp_packets:1226836 tp_drops:1206008 tp_freeze_q_cnt:25}
dropRate: 0.4957194131641815
stats: {tp_packets:1417460 tp_drops:1391830 tp_freeze_q_cnt:26}
dropRate: 0.4954383491914327
stats: {tp_packets:1603186 tp_drops:1577556 tp_freeze_q_cnt:26}
dropRate: 0.49597106587079365
stats: {tp_packets:1793263 tp_drops:1767633 tp_freeze_q_cnt:26}
dropRate: 0.4964011866676252
stats: {tp_packets:1984062 tp_drops:1951361 tp_freeze_q_cnt:27}
dropRate: 0.49584530049247566
stats: {tp_packets:2169696 tp_drops:2136995 tp_freeze_q_cnt:27}
dropRate: 0.4962034657234522
stats: {tp_packets:2358564 tp_drops:2325863 tp_freeze_q_cnt:27}
dropRate: 0.49650960512352954
stats: {tp_packets:2545841 tp_drops:2508242 tp_freeze_q_cnt:28}
dropRate: 0.49628033413776546
stats: {tp_packets:2742201 tp_drops:2694944 tp_freeze_q_cnt:30}
dropRate: 0.4956542450127778
stats: {tp_packets:2927471 tp_drops:2880214 tp_freeze_q_cnt:30}
dropRate: 0.49593151143700115
stats: {tp_packets:3117222 tp_drops:3069965 tp_freeze_q_cnt:30}
dropRate: 0.4961810593408604
stats: {tp_packets:3304548 tp_drops:3250547 tp_freeze_q_cnt:31}
dropRate: 0.49588099028313093
stats: {tp_packets:3490874 tp_drops:3436873 tp_freeze_q_cnt:31}
dropRate: 0.49610255686300325
stats: {tp_packets:3680596 tp_drops:3626595 tp_freeze_q_cnt:31}
dropRate: 0.49630494125581226
stats: {tp_packets:3872741 tp_drops:3814173 tp_freeze_q_cnt:32}
dropRate: 0.496190408790836
^C[2022-05-08T16:05:35.639+0300] INFO  ▶ Exiting ▶ [875154 main.go:82 main]
[2022-05-08T16:05:35.639+0300] INFO  ▶ Shutting down... ▶ [875119 utils.go:39 func1]
```

mertyildiran commented:
> The base images for the cross-compilation build targets below might need `linux-headers` dependency for `AF_PACKET` support:
> 
> * `--build-arg BUILDARCH=amd64 --build-arg TARGETARCH=arm64v8`
> * `--build-arg BUILDARCH=arm64v8 --build-arg TARGETARCH=amd64`
> 
> So, merge #910 first then check those build targets, only then merge this PR.

`-build-arg BUILDARCH=amd64 --build-arg TARGETARCH=arm64v8` build target works.

mertyildiran commented:
I have doubts that `AF_PACKET` works reliably because of some failures that I have experienced in the acceptance tests. Turning this PR to draft for now.

haiut commented:
based on our latest discussion - maybe we should use this PR to introduce AF_XDP ?

mertyildiran commented:
Updated PR description.

nimrod-up9 commented on tap/xdp/ipproto.c:
```diff
@@ -0,0 +1,79 @@
+#define MAX_SOCKS 64
+
+static volatile unsigned const char PROTO;
+static volatile unsigned const char PROTO = IPPROTO_ICMP;
```
Why is this `volatile` anyway? `IPPROTO_ICMP` is an enum.

mertyildiran replied:
Fixed with https://github.com/up9inc/mizu/commit/b4a9be6ad3bdf514ee3bc265fb15b921196730da

nimrod-up9 commented on tap/xdp/ipproto.c:
```diff
@@ -0,0 +1,79 @@
+	void *data_end = (void*)(long)ctx->data_end;
+	struct ethhdr *eth = data;
+	__u16 h_proto = eth->h_proto;
+	if ((void*)eth + sizeof(*eth) <= data_end) {
```
```suggestion
	if ((void*)eth + sizeof(*eth) > data_end) {
		return XDP_PASS;
	}    
```

nimrod-up9 commented on tap/xdp/ipproto.c:
```diff
@@ -0,0 +1,79 @@
+        referenced in the end BPF opcodes by file descriptor
+*/
+
+struct bpf_map_def SEC("maps") xsks_map = {
+	.type = BPF_MAP_TYPE_XSKMAP,
+	.key_size = sizeof(int),
+	.value_size = sizeof(int),
+	.max_entries = MAX_SOCKS,
+};
+
+struct bpf_map_def SEC("maps") qidconf_map = {
+	.type = BPF_MAP_TYPE_ARRAY,
+	.key_size = sizeof(int),
+	.value_size = sizeof(int),
+	.max_entries = MAX_SOCKS,
+};
```
Use the macros defined in `maps.h`.

mertyildiran replied:
Fixed with https://github.com/up9inc/mizu/commit/b4a9be6ad3bdf514ee3bc265fb15b921196730da

nimrod-up9 commented on tap/xdp/ipproto.c:
```diff
@@ -0,0 +1,79 @@
+			struct iphdr *ip = data + sizeof(*eth);
+			if ((void*)ip + sizeof(*ip) <= data_end) {
+				if (ip->protocol == PROTO) {
+					if (*qidconf)
```
```suggestion
					if (*qidconf != 0)
```

nimrod-up9 commented on tap/xdp/ipproto.c:
```diff
@@ -0,0 +1,79 @@
+			if ((void*)ip + sizeof(*ip) <= data_end) {
+				if (ip->nexthdr == PROTO) {
+					if (*qidconf)
+						return bpf_redirect_map(&xsks_map, index, 0);
```
Won't this "hijack" the packet? Mizu should be unintrusive and should not affect the way packets flow.

mertyildiran replied:
There is https://github.com/up9inc/mizu/blob/feat/afpacket/tap/source/handle_af_xdp.go#L141 which should send the packets back to the XDP socket. But upon taking another look, it might be not working as intended. What do you suggest?

nimrod-up9 replied:
It kinda goes against the main appeal of using Mizu in the first place, as opposed to other options. Except for the cpu and mem consumption, it should not affect the user's system.
By rerouting the packets we add delays and introduce the risk of dropping packets in case of an error.

I suggest copying the packet and transmitting the copy to the Go code, the same way we do it in tlsTapper.

mertyildiran replied:
Despite I searched for days or even weeks, I couldn't find a way to pass the packets to the sniffer without the `XDP_REDIRECT` involvement.

There are receive (RX) and transmit (TX) ring buffers in Linux. These buffers are shared buffers between the device driver and network interface card (NIC). According to my understanding we have to redirect the packets to an XDP socket, read the packets from that XDP socket and transmit them to somewhere such that they will continue their journey in the Linux network stack. The packets are actually captured inside [this method](https://github.com/asavie/xdp/blob/master/xdp.go#L423-L443).

Since the rebroadcast of the captured packets (transmit) is not working as intended, we might want to disable or remove `AF_XDP` from this PR.

Please also see;
- https://github.com/cloudflare/xdpcap/issues/49#issuecomment-813967235
- https://github.com/cloudflare/xdpcap/issues/32#issuecomment-563186616
- https://github.com/cloudflare/xdpcap/issues/51#issue-851220670
- https://github.com/cloudflare/xdpcap/issues/10#issuecomment-494408986

nimrod-up9 replied:
1. In tls tapper we propagate packets (actually just the tls payload) from eBPF to Go with perf buffers. Why can't we do the same here?
2. What is the benefit of XDP? Why not remove it from this PR?

nimrod-up9 reviewed (commented):
Some comments. Haven't reviewed all of the files yet.

mertyildiran commented on tap/xdp/ipproto.c:
```diff
@@ -0,0 +1,79 @@
+#define MAX_SOCKS 64
+
+static volatile unsigned const char PROTO;
+static volatile unsigned const char PROTO = IPPROTO_ICMP;
```
Fixed with https://github.com/up9inc/mizu/commit/b4a9be6ad3bdf514ee3bc265fb15b921196730da

nimrod-up9 commented on tap/source/handle_af_packet.go:
```diff
@@ -0,0 +1,151 @@
+	frameSize int, blockSize int, numBlocks int, err error) {
+
+	if snaplen < pageSize {
+		frameSize = pageSize / (pageSize / snaplen)
```
`frameSize = pageSize / (pageSize / snaplen) = snaplen`
or am I missing something because of rounding?

mertyildiran replied:
The function is copy and pasted from here: https://github.com/google/gopacket/blob/master/examples/afpacket/afpacket.go#L118-L136

```go
pageSize := os.Getpagesize()
snaplen := 127
frameSize1 := pageSize / (pageSize / snaplen)
frameSize2 := snaplen
fmt.Printf("frameSize1: %v\n", frameSize1)
fmt.Printf("frameSize2: %v\n", frameSize2)
```

There is a rounding difference:

```
frameSize1: 128
frameSize2: 127
```

I'm not sure why it's preferred like this in `gopacket`.

nimrod-up9 replied:
ok, I'll let this one go if this is how it's done in go packet. It's not important enough to stall the PR. Please just add a comment saying we took this from gopacket.

nimrod-up9 commented on tap/source/handle_af_packet.go:
```diff
@@ -0,0 +1,151 @@
+	if snaplen < pageSize {
+		frameSize = pageSize / (pageSize / snaplen)
+	} else {
+		frameSize = (snaplen/pageSize + 1) * pageSize
```
same
`frameSize = (snaplen/pageSize + 1) * pageSize = snaplen + pageSize`

mertyildiran replied:
The function is copy and pasted from here: https://github.com/google/gopacket/blob/master/examples/afpacket/afpacket.go#L118-L136

```go
pageSize := os.Getpagesize()
snaplen := 65536
frameSize1 := (snaplen/pageSize + 1) * pageSize
frameSize2 := snaplen
fmt.Printf("frameSize1: %v\n", frameSize1)
fmt.Printf("frameSize2: %v\n", frameSize2)
```

There is a rounding difference:

```
frameSize1: 69632
frameSize2: 65536
```

I'm not sure why it's preferred like this in `gopacket`.

nimrod-up9 replied:
ok, I'll let this one go if this is how it's done in go packet. It's not important enough to stall the PR. Please just add a comment saying we took this from gopacket.

nimrod-up9 commented on tap/source/handle_af_packet.go:
```diff
@@ -0,0 +1,151 @@
+func afpacketComputeSize(targetSizeMb int, snaplen int, pageSize int) (
+	frameSize int, blockSize int, numBlocks int, err error) {
+
+	if snaplen < pageSize {
```
```suggestion
	// frameSize calculation was taken from gopacket's afpacket.go
	if snaplen < pageSize {
```

nimrod-up9 commented:
@mertyildiran I approved the PR following your removal of xdp.

mertyildiran commented:
> @mertyildiran I approved the PR following your removal of xdp.

@nimrod-up9 OK, thanks a lot!
